### PR TITLE
Re-Evolution: Add functionality for battle-only formes

### PR DESCRIPTION
**Pull request by Seerd**
Re-evolution mod, /showevo and /reevo now check for the base species to not lose track of the pre-evolution after changing forms, /showevo now also shows the stats changes a battle form gives.

### CODE REVIEW

File: data/rulesets.ts [hunk 1]
```diff
@@ -1902,16 +1902,32 @@ export const Rulesets: {[k: string]: FormatData} = {
...
 		onModifySpecies(species, target) {
 			const newSpecies = this.dex.deepClone(species);
-			if (!newSpecies.prevo) return;
-			const prevoSpecies = this.dex.species.get(newSpecies.prevo);
-			let statid: StatID;
-			newSpecies.bst = 0;
-			for (statid in prevoSpecies.baseStats) {
+			const baseSpe = this.dex.species.get(species.baseSpecies);
+			if(!newSpecies.prevo)
+			{
```
Comment (KrisXV) [inline data/rulesets.ts:1907]:
```suggestion
			if (!newSpecies.prevo) {
```

**Comment by KrisXV** [comment]
Make sure you're testing your code with `npm test` while making pull requests to make sure the code matches the linter's standards

**Comment by KrisXV** [comment]
Also, I'm unsure you're approaching this properly/if the code even works? It seems very stuck together and there's not a check for `species.battleOnly`

**Comment by Seerd** [comment]
I'm not using `species.battleOnly` because some forms like Megas don't seem to have that check, that kind of made me look for a workaround and that's why it ended longer than what I was expecting, the code works properly on my local server at least, also just made the changes from the npm test, sorry for the inconvenience.

**Comment by KrisXV** [comment]
Make sure you're running `npm test`, there's still a lot of linter errors

**Comment by KrisXV** [comment]
Also, Mega formes do have `battleOnly`, it's just assigned outside of `pokedex.ts`

**Comment by Seerd** [comment]
It should work fine now.

File: server/chat-plugins/othermetas.ts [hunk 2]
```diff
@@ -620,46 +620,93 @@ export const commands: Chat.ChatCommands = {
...
-		if (!isReEvo) {
-			deltas.tier = 'CE';
-			deltas.weightkg = evo.weightkg - prevoSpecies.weightkg;
-			deltas.types = [];
-			if (evo.types[0] !== prevoSpecies.types[0]) deltas.types[0] = evo.types[0];
-			if (evo.types[1] !== prevoSpecies.types[1]) {
-				deltas.types[1] = evo.types[1] || evo.types[0];
+			const bevo = Dex.species.get(evo.baseSpecies);
+			if (!bevo.prevo) { throw new Chat.ErrorMessage(`Error: ${evo.name} is not an evolution.`); }
+			const prevoSpecies = Dex.species.get(bevo.prevo);
```
Comment (KrisXV) [inline server/chat-plugins/othermetas.ts:625]:
```suggestion
			const evoBaseSpecies = Dex.species.get(evo.baseSpecies);
			if (!evoBaseSpecies.prevo) throw new Chat.ErrorMessage(`Error: ${evoBaseSpecies.name} is not an evolution.`);
			const prevoSpecies = Dex.species.get(evoBaseSpecies.prevo);
```

File: server/chat-plugins/othermetas.ts [hunk 3]
```diff
@@ -620,46 +620,93 @@ export const commands: Chat.ChatCommands = {
...
+					if (deltas.types[0] === deltas.types[1]) deltas.types = [deltas.types[0]];
+			    } else {
+					deltas.types = null;
+			    }
+		    }
+		    deltas.bst = 0;
+		    let i: StatID;
+		    for (i in evo.baseStats) {
+				const statChange = bevo.baseStats[i] - prevoSpecies.baseStats[i];
+				const frmChange = evo.baseStats[i] - bevo.baseStats[i];
```
Comment (KrisXV) [inline server/chat-plugins/othermetas.ts:648]:
```suggestion
				const statChange = evoBaseSpecies.baseStats[i] - prevoSpecies.baseStats[i];
				const formeChange = evo.baseStats[i] - evoBaseSpecies.baseStats[i];
```

File: server/chat-plugins/othermetas.ts [hunk 4]
```diff
@@ -620,46 +620,93 @@ export const commands: Chat.ChatCommands = {
...
+				if (!isReEvo) {
+			    if (!evo.prevo) {
+						deltas.baseStats[i] = frmChange;
```
Comment (KrisXV) [inline server/chat-plugins/othermetas.ts:651]:
```suggestion
						deltas.baseStats[i] = formeChange;
```

File: server/chat-plugins/othermetas.ts [hunk 5]
```diff
@@ -620,46 +620,93 @@ export const commands: Chat.ChatCommands = {
...
+					} else {
+						deltas.baseStats[i] = statChange;
+					}
+				} else {
+					deltas.baseStats[i] = Utils.clampIntRange(bevo.baseStats[i] + statChange, 1, 255);
```
Comment (KrisXV) [inline server/chat-plugins/othermetas.ts:656]:
```suggestion
					deltas.baseStats[i] = Utils.clampIntRange(evoBaseSpecies.baseStats[i] + statChange, 1, 255);
```

File: server/chat-plugins/othermetas.ts [hunk 6]
```diff
@@ -620,46 +620,93 @@ export const commands: Chat.ChatCommands = {
...
+		    if (!isReEvo) {
+			    deltas.tier = 'CE';
+			    deltas.weightkg = evo.weightkg - prevoSpecies.weightkg;
+			    deltas.types = [];
+			    if (evo.types[0] !== prevoSpecies.types[0]) deltas.types[0] = evo.types[0];
+			    if (evo.types[1] !== prevoSpecies.types[1]) {
+				    deltas.types[1] = evo.types[1] || evo.types[0];
+			    }
+			    if (deltas.types.length) {
 				// Undefined type remover
```
Comment (KrisXV) [inline server/chat-plugins/othermetas.ts:636]:
```suggestion
					// Undefined type remover
```

**Pull request merged**